clippy_lints/src/assigning_clones.rs

```diff
@@ -0,0 +1,204 @@
+    ///
+    /// Note that this only provides any actual improvement if the type has explicitly implemented
+    /// the `clone_from()` trait method, since the trait-provided implementation will just call
+    /// `clone()`.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[derive(Clone)]
+    /// struct Thing;
+    ///
+    /// pub fn assign_to_ref(a: &mut Thing, b: Thing) {
+    ///     *a = b.clone();
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// #[derive(Clone)]
+    /// struct Thing;
+    ///
+    /// pub fn assign_to_ref(a: &mut Thing, b: Thing) {
+    ///     a.clone_from(&b);
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub ASSIGNING_CLONES,
+    perf,
+    "assigning the result of cloning may be inefficient"
+}
+declare_lint_pass!(AssigningClones => [ASSIGNING_CLONES]);
+
+impl<'tcx> LateLintPass<'tcx> for AssigningClones {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, assign_expr: &'tcx hir::Expr<'_>) {
+        let ExprKind::Assign(assign_target, clone_call, _span) = assign_expr.kind else { return };
+        // TODO: Also look for `Clone::clone` function calls, not just method calls
+        let ExprKind::MethodCall(method_name, clone_receiver, args, _span) = clone_call.kind else { return };
+
+        // Fast syntactic check: if it has args it can't be the call we are looking for,
+        // so we don't even need to consult the types.
+        if !args.is_empty() {
+            return;
+        }
+
+        let op = if method_name.ident.name == sym::clone {
+            Op::Clone
+        } else if method_name.ident.name == sym!(to_owned) {
+            Op::ToOwned
+        } else {
+            return;
+        };
+
+        if ok_to_suggest(cx, op, assign_target, clone_call).is_some() {
+            suggest(cx, op, assign_expr, assign_target, clone_receiver);
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum Op {
+    Clone,
+    ToOwned,
+}
+
+// Return `Some(())` iff we have confirmed that the call is in fact one we want to replace.
+fn ok_to_suggest<'tcx>(
+    cx: &LateContext<'tcx>,
+    op: Op,
+    assign_target: &hir::Expr<'tcx>,
+    clone_call: &hir::Expr<'tcx>,
+) -> Option<()> {
+    // Check that the call is actually a call to the trait.
+    // TODO: Actually we are currently just checking that the result of the call is
+    // a type that implements the trait, which is a bad proxy for it.
+    let clone_result_type = cx.typeck_results().expr_ty_adjusted(clone_call);
+    if !(implements_trait(cx, clone_result_type, op.expected_trait(cx)?, &[])) {
+        return None;
+    }
+
+    // If we're assigning to a dereferenced reference, then we know the place is already valid.
+    // On the other hand, if the place is a variable or a Box, it might be uninitialized,
+    // in which case the suggestion might be wrong.
+    // TODO: Actually ask whether the place is uninitialized at this point, instead of
+    // guessing based on the syntax and type.
```

Comment:
Finding the origin of a variable can be a bit difficult, when it comes to patterns, function parameters etc. Though, in this case, we only want to find the `let <var>` statement if it's a local value. This is a rough sketch of how to check for this:

```
if let Some(id) = path_to_local(expr) {
    if let Some(Node::Stmt(stmt)) = cx.tcx.hir().find(id) {
        // if stmt is let, check for init expr
    }
}
```

[`clippy_utils::expr_or_init`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.expr_or_init.html) could also be interesting to look at :)